### PR TITLE
VZ-11651: Analysis tool redacted values mapping

### PR DESCRIPTION
**Pull request by marcogelle**
This PR adds functionality to the analysis tool to create a mapping from redacted values to their original ones, stored in a file called `sensitive-do-not-share-redaction-map.csv`. However, this PR only provides this functionality as helper functions that are not actually called yet.

This PR also makes some changes under `tests/e2e/verify-analyze-tool` to address a test failure/bug.

### CODE REVIEW

File: tools/vz/pkg/helpers/vzsanitize.go [hunk 1]
```diff
@@ -89,3 +124,11 @@ func getSha256Hash(line string) string {
 	hexString := hex.EncodeToString(hashedVal[:])
 	return hexString
 }
+
+// determineRedactedValuesMap returns the what map of redacted values to use, according to the override provided
```
Comment (zabdulre) [inline tools/vz/pkg/helpers/vzsanitize.go:128]: Typo, remove "what"

File: tools/vz/pkg/helpers/vzsanitize.go [hunk 2]
```diff
@@ -58,28 +65,56 @@ func SanitizeString(l string) string {
...
+		if err = csvWriter.Write([]string{r, s}); err != nil {
+			LogError(fmt.Sprintf("An error occurred while writing the file %s: %s\n", fileName, err.Error()))
+			return err
+		}
+	}
+	csvWriter.Flush()
+	return nil
+}
+
+func (rp regexPlan) compilePlan(redactedValuesOverride map[string]string) func(string) string {
```
Comment (zabdulre) [inline tools/vz/pkg/helpers/vzsanitize.go:95]: Is the override required? Have you considered using the global map for testing purposes?

File: tools/vz/pkg/helpers/vzsanitize.go [hunk 3]
```diff
@@ -58,28 +65,56 @@ func SanitizeString(l string) string {
...
-		l = regexp.MustCompile(eachRegex.regex).ReplaceAllStringFunc(l, eachRegex.compilePlan())
+		l = regexp.MustCompile(eachRegex.regex).ReplaceAllStringFunc(l, eachRegex.compilePlan(redactedValuesOverride))
 	}
 	return l
 }
 
-func (rp regexPlan) compilePlan() func(string) string {
+// WriteRedactionMapFile creates a CSV file to document all the values this tool has
+// redacted so far, stored in the redactedValues map.
+func WriteRedactionMapFile(captureDir string, redactedValuesOverride map[string]string) error {
```
Comment (bobdonat) [inline tools/vz/pkg/helpers/vzsanitize.go:75]: Why is this file needed?

File: tools/vz/pkg/helpers/vzsanitize.go (hunk 1 above)
Comment (marcogelle) [reply tools/vz/pkg/helpers/vzsanitize.go:128]: will do

File: tools/vz/pkg/helpers/vzsanitize.go (hunk 2 above)
Comment (zabdulre) [reply tools/vz/pkg/helpers/vzsanitize.go:95]: Ok, sounds good!

File: tools/vz/pkg/constants/constants.go [hunk 4]
```diff
@@ -170,6 +170,9 @@ const (
 	BugReportError   = "ERROR: The bug report noticed one or more issues while capturing the resources. Please go through error(s) in the standard error."
 	BugReportWarning = "WARNING: Please examine the contents of the bug report for any sensitive data"
 
+	// File containing a map from redacted values to their original values
+	RedactionMap = "redaction-map.csv"
```
Comment (agv426) [inline tools/vz/pkg/constants/constants.go:174]: Thinking we want the name of this to make it clear they should not share this.

maybe we can call it something like "sensitive-do-not-share-redaction-map.csv

Comment (marcogelle) [reply tools/vz/pkg/constants/constants.go:174]: Ok, I'll go ahead with renaming it to `sensitive-do-not-share-redaction-map.csv`.

File: tools/vz/pkg/helpers/vzsanitize.go [hunk 5]
```diff
@@ -58,28 +65,56 @@ func SanitizeString(l string) string {
 	}
 	knownHostNamesMutex.Unlock()
 	for _, eachRegex := range regexToReplacementList {
-		l = regexp.MustCompile(eachRegex.regex).ReplaceAllStringFunc(l, eachRegex.compilePlan())
+		l = regexp.MustCompile(eachRegex.regex).ReplaceAllStringFunc(l, eachRegex.compilePlan(redactedValuesOverride))
```
Comment (bhatpmk) [inline tools/vz/pkg/helpers/vzsanitize.go:68]: I think we need a lock here as well. I got a fatal error: concurrent map read and map write, once when I run the CLI with this change.

Comment (marcogelle) [reply tools/vz/pkg/helpers/vzsanitize.go:68]: Strange, I've run the CLI with these changes many times without seeing that error.
However, locking there makes sense to me, I'll work on adding it now.

File: tools/vz/pkg/helpers/vzsanitize_test.go [hunk 6]
```diff
@@ -39,3 +43,80 @@ func TestSanitizeALine(t *testing.T) {
...
+	testRedactedValues := make(map[string]string)
+	// redact a variety of inputs, as well as inputting a value more than once.
+	testInputs := []string{testIP, testOCID, testSSH, testUserData, testOPCID, testIP}
+	for _, input := range testInputs {
+		redact(input, testRedactedValues)
+	}
+	numUniqueInputs := 5
+	a.Len(testRedactedValues, numUniqueInputs)
+
+	// write the redacted values to /tmp/redaction-map.csv
```
Comment (bhatpmk) [inline tools/vz/pkg/helpers/vzsanitize_test.go:70]: The name of the file will be sensitive-do-not-share-redaction-map.csv

File: tools/vz/pkg/helpers/vzsanitize_test.go [hunk 7]
```diff
@@ -39,3 +43,80 @@ func TestSanitizeALine(t *testing.T) {
 	strictCheck(testUserData, testUserDataToRemove)
 	strictCheck(testOPCID, testOPCIDToRemove)
 }
+
+func TestWriteRedactionMapFileEmpty(t *testing.T) {
```
Comment (bhatpmk) [inline tools/vz/pkg/helpers/vzsanitize_test.go:47]: Add comments for the newly added functions (for the existing one as well, if possible) as recommended at https://tip.golang.org/doc/comment

File: tools/vz/pkg/helpers/vzsanitize_test.go [hunk 8]
```diff
@@ -39,3 +43,80 @@ func TestSanitizeALine(t *testing.T) {
...
+	a := assert.New(t)
+	testRedactedValues := make(map[string]string)
+	err := WriteRedactionMapFile("/tmp", testRedactedValues)
```
Comment (bhatpmk) [inline tools/vz/pkg/helpers/vzsanitize_test.go:50]: Instead of /tmp, os.TempDir() can be used.

File: tools/vz/pkg/helpers/vzsanitize_test.go [hunk 9]
```diff
@@ -39,3 +43,80 @@ func TestSanitizeALine(t *testing.T) {
...
+	a.Equal(numUniqueInputs, numLines)
+}
+
+func TestRedact(t *testing.T) {
+	a := assert.New(t)
+	testRedactedValues := make(map[string]string)
+
+	// test that redacting the same value repeatedly returns the same value
+	redactedIP := redact(testIP, testRedactedValues)
+	a.Contains(redactedIP, "REDACTED-")
```
Comment (bhatpmk) [inline tools/vz/pkg/helpers/vzsanitize_test.go:106]: "REDACTED-" can be a constant defined in pkg/constants/constants.go and referred from the test and vzsanitize.go

File: tools/vz/pkg/helpers/vzsanitize_test.go [hunk 10]
```diff
@@ -39,3 +43,80 @@ func TestSanitizeALine(t *testing.T) {
...
+	// redact a variety of inputs, as well as inputting a value more than once.
+	testInputs := []string{testIP, testOCID, testSSH, testUserData, testOPCID, testIP}
+	for _, input := range testInputs {
+		redact(input, testRedactedValues)
+	}
+	numUniqueInputs := 5
+	a.Len(testRedactedValues, numUniqueInputs)
+
+	// write the redacted values to /tmp/redaction-map.csv
+	err := WriteRedactionMapFile("/tmp", testRedactedValues)
```
Comment (Abhijithang) [inline tools/vz/pkg/helpers/vzsanitize_test.go:71]: Please move the file name to constant when used multiple times. Same applies to any hard coded values used multiple times.

File: tools/vz/pkg/helpers/vzsanitize_test.go (hunk 6 above)
Comment (marcogelle) [reply tools/vz/pkg/helpers/vzsanitize_test.go:70]: Ok, will fix this

File: tools/vz/pkg/helpers/vzsanitize_test.go (hunk 7 above)
Comment (marcogelle) [reply tools/vz/pkg/helpers/vzsanitize_test.go:47]: Ok, will do

File: tools/vz/pkg/helpers/vzsanitize_test.go (hunk 10 above)
Comment (marcogelle) [reply tools/vz/pkg/helpers/vzsanitize_test.go:71]: Ok, will do

**Review by bhatpmk — APPROVED** [review]

**Review by Abhijithang — APPROVED** [review]

**Review by agv426 — APPROVED** [review]
